lib/Sema/TypeCheckAttr.cpp

```diff
@@ -2799,6 +2852,285 @@ void AttributeChecker::visitDifferentiableAttr(DifferentiableAttr *attr) {
   }
 }
 
+// SWIFT_ENABLE_TENSORFLOW
+// Makes a function with the same generic signature and extinfo as `copy`, but
+// with `params` parameters and `retTy` return type.
+static AnyFunctionType *
+makeFunctionType(AnyFunctionType *copy, ArrayRef<AnyFunctionType::Param> params,
+                 Type retTy, GenericSignature *genericSignature) {
+  if (!genericSignature)
+    if (auto *genericFunctionType = copy->getAs<GenericFunctionType>())
+      genericSignature = genericFunctionType->getGenericSignature();
+  if (genericSignature)
+    return GenericFunctionType::get(genericSignature, params, retTy,
+                                    copy->getExtInfo());
+  return FunctionType::get(params, retTy, copy->getExtInfo());
+}
+
+// SWIFT_ENABLE_TENSORFLOW
+// Return the original function type corresponding to the given derivative
+// function type.
+static AnyFunctionType *
+getAutoDiffOriginalFunctionType(AnyFunctionType *derivativeType) {
+  // Unwrap curry levels.
+  SmallVector<AnyFunctionType *, 2> curryLevels;
+  auto *currentLevel = derivativeType;
+  while (currentLevel != nullptr) {
+    curryLevels.push_back(currentLevel);
+    currentLevel = currentLevel->getResult()->getAs<AnyFunctionType>();
+  }
+
+  auto derivativeResult = curryLevels.back()->getResult()->getAs<TupleType>();
+  assert(derivativeResult && derivativeResult->getNumElements() == 2 &&
+         "Expected derivative result to be a two-element tuple");
+  auto originalResult = derivativeResult->getElement(0).getType();
+  auto genericSignature = derivativeType->getOptGenericSignature();
+  auto *originalType = makeFunctionType(
+      curryLevels.back(), curryLevels.back()->getParams(), originalResult,
+      curryLevels.size() == 1 ? genericSignature : nullptr);
+
+  // Wrap the associated function type in additional curry levels.
+  auto curryLevelsWithoutLast =
+  ArrayRef<AnyFunctionType *>(curryLevels).drop_back(1);
```

Comment:
```suggestion
      ArrayRef<AnyFunctionType *>(curryLevels).drop_back(1);
```

Reply:
Fixed in https://github.com/apple/swift/pull/22225/commits/6c5a209cfecc95609e088a5e3a165631faa0768e.